server/src/main/java/io/druid/server/coordination/SegmentChangeRequestHistory.java

```diff
@@ -0,0 +1,300 @@
+  {
+    this.maxSize = maxSize;
+    this.changes = new ArrayList<>(maxSize);
+
+    this.waitingFutures = new LinkedHashMap<>();
+
+    this.resolveWaitingFuturesRunnable = new Runnable()
+    {
+      @Override
+      public void run()
+      {
+        resolveWaitingFutures();
+      }
+    };
+
+    this.singleThreadedExecutor = Executors.newSingleThreadExecutor(
+        new ThreadFactoryBuilder().setDaemon(true)
+                                  .setNameFormat(
+                                      "SegmentChangeRequestHistory"
+                                  )
+                                  .build()
+    );
+  }
+
+
+
+  /**
+   * Add batch of segment changes update.
+   */
+  public synchronized void addSegmentChangeRequests(List<DataSegmentChangeRequest> requests)
+  {
+    for (DataSegmentChangeRequest request : requests) {
+      if (changes.size() >= maxSize) {
+        changes.remove(0);
+      }
+      changes.add(new Holder(request, getLastCounter().inc()));
+    }
+
+    singleThreadedExecutor.execute(resolveWaitingFuturesRunnable);
+  }
+
+  /**
+   * Add single segment change update.
+   */
+  public synchronized void addSegmentChangeRequest(DataSegmentChangeRequest request)
+  {
+    addSegmentChangeRequests(ImmutableList.of(request));
+  }
+
+  /**
+   * Returns a Future that , on completion, returns list of segment updates and associated counter.
+   * If there are no update since given counter then Future completion waits till an updates is provided.
+   *
+   * If counter is very older than max number of changes maintained than IAE exception is thrown.
+   *
+   * If there were no updates to provide immediately then a future is created and returned to caller. This future
+   * is added to the "waitingFutures" list and all the futures in the list get resolved as soon as a segment
+   * update is provided.
+   */
+  public synchronized ListenableFuture<SegmentChangeRequestsSnapshot> getRequestsSince(final Counter counter)
+  {
+    final CustomSettableFuture future = new CustomSettableFuture(waitingFutures);
+
+    if (counter.counter < 0) {
+      future.setException(new IAE("counter must be >= 0"));
+      return future;
+    }
+
+    Counter lastCounter = getLastCounter();
+
+    if (counter.counter == lastCounter.counter) {
+      if (!counter.matches(lastCounter)) {
+        future.setException(new IAE("counter failed to match"));
+      } else {
+        synchronized (waitingFutures) {
+          waitingFutures.put(future, counter);
+        }
+      }
+    } else {
+      try {
+        future.set(getRequestsSinceWithoutWait(counter));
+      } catch (Exception ex) {
+        future.setException(ex);
+      }
+    }
+
+    return future;
+  }
+
+  private synchronized SegmentChangeRequestsSnapshot getRequestsSinceWithoutWait(final Counter counter)
+  {
+    Counter lastCounter = getLastCounter();
+
+    if (counter.counter >= lastCounter.counter) {
+      throw new IAE("counter >= last counter");
```

Comment:
Please include the actual values that came in here, just the message as it exists won't really help much with debugging if it actually gets thrown.